data/blog/2025/2025-08-06-strangler-fig-pattern.md

```diff
@@ -0,0 +1,139 @@
+Bricks made of gold (OK, that was a bit much).
+A well-implemented Strangler Fig Pattern will have that wall provide the same functionality throughout its rebuilding process.
+At the end of the migration, the whole wall will contain all-new bricks, built better than before, anxiously awaiting new features.
+
+## Why Not Rewrite the Old Platform Instead?
+
+Many organizations are weighed down with legacy applications that:
+
+- Are a vital business component that can't be put on pause during additional development.
+- Are difficult to change or scale because of platform limitations or sheer complexity.
+- Use outdated or discontinued technology stacks.
+
+One of the options for companies that need to improve their workflow is a **complete rewrite**, but it carries several **disadvantages**:
+
+- There is a risk of not migrating important domain logic and business logic that is not well documented.
+- It's expensive and time-consuming to rewrite code that has been under constant improvement over many years.
+- Likely to fail, especially because of learning curves and hiring challenges for developers of the new architecture.
+- It's often rejected by stakeholders because of no perceived benefits or short-term value.
+
+## What Are the advantages of the Strangler Fig Pattern?
+
+The **Strangler Fig Pattern** offers a **safer, controlled way** to:
+
+- **Incremental modernization** based on a detailed plan.
+- **Faster delivery of features** with short coding sprints, aimed at replacing isolated components.
+- **Improved maintainability** for each new component because of its decoupled state, which enables separate testing.
+- **Flexible development pacing** where you set your goals in terms of development schedule (the process is long and can be interrupted more easily) and allocated resources (the developers who actually work on the migration).
+- **Mitigate risks** by avoiding a full rewrite in favor of small, manageable changes.
+
+## How Strangler Fig Pattern Works: Step-by-Step
+
+### Step 1: Understand and Isolate Functionality
+
+Analyze the legacy system to identify well-defined boundaries or modules.
+Select a coherent part of the functionality that can be isolated and replaced independently (e.g. user authentication, payment processing).
+It's a good idea to work with the original developers for this step, if they are still available.
+
+### Step 2: Build the Replacement Code
+
+Implement the isolated functionality in the new platform using modern architectures (e.g. microservices, middleware).
+Ensure that the new service meets all the functional and non-functional requirements (performance, security etc.).
+
+### Step 3: Intercept Calls Via the Routing Layer
+
+Use a routing mechanism (like an API gateway, proxy, service mesh) to redirect calls that use that functionality to the new service.
+This allows the rest of the application to continue working as-is while the new component handles specific operations.
```

Comment:
Or even just the web server...

Also, I'm wondering if authentication/authorization should be the example in the previous section, as that likely has deeper ties within the application, and it will be hard to "route" that sort of functionality.

And that's where this all gets more complicated: because the new architecture may also introduce a new process/workflow for implementing fundamental building blocks like this.

Choose a simpler example ("contact form", "product gallery", etc.), and make a note of the complexity when dealing with things like authn/authz that will likely require bigger, system-wide changes.